README.md
# A Penrose tiling

This project generates a SVG file with a *Penrose tiling* (an aperiodic tiling of the plane with two tiles), using the [cairo-fortran](https://github.com/vmagnin/cairo-fortran) bindings as a fpm (Fortran Package Manager) dependency.

The algorithm uses acute and obtuse golden triangles, with two interlaced recursive procedures. See the figures in https://fr.wikipedia.org/wiki/Pavage_de_Penrose to understand.

See the project https://github.com/vmagnin/hat_polykite for more information on the *ein Stein* problem and the tiling of the plane, and the most recent results in this domain.

![A Penrose tiling](figures/penrose_tiling_10.svg)

Be careful with the number of recursions: the SVG file is growing like the Fibonacci sequence! 12 recursions will generate a 12.7 Mio SVG file. 15 recursions will give a 228 Mio file and your SVG viewer may freeze or crash...

## Requirements and dependencies

You need:

* a modern Fortran compiler, for example GFortran or the Intel ifort/ifx compilers. See the [Fortran-lang.org compilers page](https://fortran-lang.org/compilers/) for other compilers.
* The Cairo development files (`libcairo2-dev` package in Ubuntu).
* The Fortran Package Manager [fpm](https://fpm.fortran-lang.org/).

## Running the program

Just type `fpm run` and fpm will manage the dependencies and build and run the program:

```bash
$ fpm run
 + mkdir -p build/dependencies
Initialized empty Git repository in penrose_tiling/build/dependencies/cairo-fortran/.git/
remote: Enumerating objects: 23, done.
remote: Counting objects: 100% (23/23), done.
remote: Compressing objects: 100% (22/22), done.
remote: Total 23 (delta 0), reused 14 (delta 0), pack-reused 0
Unpacking objects: 100% (23/23), 82.99 KiB | 1.51 MiB/s, done.
From https://github.com/vmagnin/cairo-fortran
 * branch            HEAD       -> FETCH_HEAD
cairo-enums.f90                        done.
cairo-auto.f90                         done.
triangles.f90                          done.
libpenrose_tiling.a                    done.
main.f90                               done.
penrose_tiling                         done.
[100%] Project compiled successfully.
 ----------------------------------
           8 recursions
 These are Fibonacci numbers:
         610 blue acute triangles
         377 yellow obtuse triangles
 ----------------------------------
 Output file: penrose_tiling.svg
 ----------------------------------
```


## Licenses

This project is licensed under the [GNU General Public License version 3 or later](http://www.gnu.org/licenses/gpl.html).

The documentation is under the [GNU Free Documentation License 1.3 or any later version](http://www.gnu.org/licenses/fdl.html). The figures are under [CC-BY-NC-SA 4.0](https://creativecommons.org/licenses/by-nc-sa/4.0/) license.

## Cairo documentation

* https://cairographics.org/documentation/
* API Reference Manual: https://cairographics.org/manual/

## Bibliography

* Hargittai, Istvan, et Balazs Hargittai. « 2020 Physics Nobel Laureate Roger Penrose and the Penrose Pattern as a Forerunner of Generalized Crystallography ». *Structural Chemistry* 32, no.1 (1st February 2021): 1‑7. https://doi.org/10.1007/s11224-020-01669-8
* https://en.wikipedia.org/wiki/Einstein_problem
* https://en.wikipedia.org/wiki/Penrose_tiling
* https://fr.wikipedia.org/wiki/Pavage_de_Penrose
* https://en.wikipedia.org/wiki/Golden_ratio
* https://en.wikipedia.org/wiki/Golden_triangle_(mathematics)


CHANGELOG.md
# Changelog
All notable changes to the gtk-fortran project are documented in this file. The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/).

## [penrose_tiling 1.0] 2023-06-04

* Fortran version, using cairo-fortran.


src/triangles.f90
! This file is part of the penrose_tiling Fortran project.
! Copyright (C) 2023 Vincent MAGNIN
!
! This is free software; you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation; either version 3, or (at your option)
! any later version.
!
! This software is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License along with
! this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
! If not, see <http://www.gnu.org/licenses/>.
!------------------------------------------------------------------------------
! Contributed by Vincent Magnin
! GFA BASIC version: 1991-09-12
! Java version: 2002-12-14
! Fortran version: 2023-06-01
! Last modifications: 2023-06-05
!------------------------------------------------------------------------------

module triangles
    ! Cairo uses C doubles:
    use, intrinsic :: iso_c_binding, only: dp=>c_double, c_ptr
    ! Use the cairo-fortran bindings:
    use cairo

    implicit none
    private
    public :: phi, acute_triangle, nb_acute, nb_obtuse

    ! Golden ratio:
    real(dp), parameter :: phi = (1 + sqrt(5._dp)) / 2
    ! Triangles counters:
    integer :: nb_acute  = 0
    integer :: nb_obtuse = 0

contains

    subroutine draw_triangle(P1, P2, P3, r, g, b, cr)
        complex(dp), intent(in) :: P1, P2, P3
        real(dp), intent(in)    :: r, g, b
        ! Cairo context:
        type(c_ptr), intent(in) :: cr

        call cairo_set_source_rgb(cr, r, g, b)

        ! Path of the triangle:
        call cairo_move_to(cr, P1%re, P1%im)
        call cairo_line_to(cr, P2%re, P2%im)
        call cairo_line_to(cr, P3%re, P3%im)
        call cairo_close_path(cr)

        ! Fill the triangle:
        call cairo_fill_preserve(cr)
        ! and draw its edges in black:
        call cairo_set_source_rgb(cr, 0._dp, 0._dp, 0._dp)
        call cairo_stroke(cr)
    end subroutine draw_triangle


    recursive subroutine acute_triangle(P1, P2, P3, n, nmax, cr)
        complex(dp), intent(in) :: P1, P2, P3
        integer, intent(in)     :: n, nmax
        type(c_ptr), intent(in) :: cr
        complex(dp) :: PP1, PP2, PP3, PP4, PP5
        real(dp) :: d1, d2, d3

        if (n < nmax) then  ! *** Recursion case ***
            ! Lengths of the sides of the triangle:
            d1 = abs(P1 - P2)
            d2 = abs(P1 - P3)
            d3 = abs(P2 - P3)

            if ((d1 <= d2) .and. (d1 <= d3)) then
                PP1 = P3
                PP2 = P2
                PP3 = P1
            else if ((d2 <= d1) .and. (d2 <= d3)) then
                PP1 = P2
                PP2 = P1
                PP3 = P3
            else
                PP1 = P1
                PP2 = P2
                PP3 = P3
            end if

            PP4 = (phi*PP3 + PP1) / (1 + phi)
            PP5 = (phi*PP1 + PP2) / (1 + phi)

            call acute_triangle (PP4, PP5, PP2, n+1, nmax,  cr)
            call acute_triangle (PP2, PP3, PP4, n+1, nmax,  cr)
            call obtuse_triangle(PP1, PP4, PP5, n+1, nmax,  cr)
        else                ! *** Terminating case ***
            nb_acute = nb_acute + 1
            ! Draw the triangle with a blue:
            call draw_triangle(P1, P2, P3, 0._dp, 0.392_dp, 1._dp, cr)
        end if
    end subroutine acute_triangle


    recursive subroutine obtuse_triangle(P1, P2, P3, n, nmax, cr)
        complex(dp), intent(in) :: P1, P2, P3
        integer, intent(in)     :: n, nmax
        type(c_ptr), intent(in) :: cr
        complex(dp) :: PP1, PP2, PP3, PP4
        real(dp) :: d1, d2, d3

        if (n < nmax) then  ! *** Recursion case ***
            ! Sides lengths of the triangle:
            d1 = abs(P1 - P2)
            d2 = abs(P1 - P3)
            d3 = abs(P2 - P3)

            if (((d1 >= d2) .and. (d1 >= d3))) then
                PP1 = P3
                PP2 = P2
                PP3 = P1
            else if ((d2 >= d1) .and. (d2 >= d3)) then
                PP1 = P2
                PP2 = P1
                PP3 = P3
            else
                PP1 = P1
                PP2 = P2
                PP3 = P3
            end if

            PP4 = (phi*PP2 + PP3) / (1 + phi)

            call acute_triangle (PP1, PP3, PP4, n+1, nmax,  cr)
            call obtuse_triangle(PP1, PP2, PP4, n+1, nmax,  cr)
        else                ! *** Terminating case ***
            nb_obtuse = nb_obtuse + 1
            ! Draw the triangle in yellow:
            call draw_triangle(P1, P2, P3, 1._dp, 1._dp, 0._dp, cr)
        end if
    end subroutine obtuse_triangle

end module triangles


other_languages/penrose.java
// This file is part of the penrose_tiling Fortran project.
// Copyright (C) 2023 Vincent MAGNIN
//
// This is free software; you can redistribute it and/or modify
// it under the terms of the GNU General Public License as published by
// the Free Software Foundation; either version 3, or (at your option)
// any later version.
//
// This software is distributed in the hope that it will be useful,
// but WITHOUT ANY WARRANTY; without even the implied warranty of
// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
// GNU General Public License for more details.
//
// You should have received a copy of the GNU General Public License along with
// this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
// If not, see <http://www.gnu.org/licenses/>.
//------------------------------------------------------------------------------
// Contributed by Vincent Magnin
// Java version: 2002-12-14
// Last modifications: 2023-06-05
//------------------------------------------------------------------------------

/* Pavage de Penrose avec triangles d'or (pavage de type 0)
   Vincent MAGNIN
   date de création (GFA BASIC) : 12 septembre 1991
   en java : 14 décembre 2002
   Dernières modifications :
     - 12-03-2018 : javac 1.8.0_151
     - 05-06-2023 : amélioration de la présentation
   https://fr.wikipedia.org/wiki/Pavage_de_Penrose
   $ javac penrose.java && java -cp . penrose
*/

import java.awt.* ;
import javax.swing.* ;

public class penrose
{

  public static void main (String args[])
    {
      maFenetre fenetre= new maFenetre() ;
      fenetre.setVisible(true) ;
    }
}


class maFenetre extends JFrame {
   private Panneau pano ;
   public maFenetre()
         {
      setTitle("Un pavage de Penrose (a Penrose tiling)") ;
          setSize(1600, 1024) ;
          pano = new Panneau() ;
          getContentPane().add(pano) ;
          setDefaultCloseOperation(JFrame.EXIT_ON_CLOSE) ;
          }
}

class Panneau extends JPanel
{
  double or ;
  long niveau ;
  Color jaune  = new Color (255, 255, 0) ;
  Color noir   = new Color (0, 0, 0) ;
  Color bleu   = new Color (0, 100, 255) ;

  public void paintComponent (Graphics g)
  {
    super.paintComponent(g) ;

    or = (1.0 + Math.sqrt(5)) / 2.0 ;  //Nombre d'or
    niveau = 9 ; // on peut aller jusque 10 environ
    double d = 960 ;
    double h = d*Math.sqrt(or*or - 1.0/4.0) ;

    acute_triangle(0, 0 , 0, d, h, d/2.0, 1, g) ;
  }

  void acute_triangle(double x1, double y1, double x2, double y2, double x3, double y3, long n, Graphics g)
  {
    double x4, y4, x5, y5 ;
    double d1, d2, d3 ;
    double inter ;

    if (n < niveau) {
      d1 = distance(x1, y1, x2, y2) ;
      d2 = distance(x1, y1, x3, y3) ;
      d3 = distance(x2, y2, x3, y3) ;

      n++ ;

      if (MIN(d1, d2, d3) == d1) {
        inter = x1 ;
        x1 = x3 ;
        x3 = inter ;
        inter = y1 ;
        y1 = y3 ;
        y3 = inter ;
      }
      else if (MIN(d1, d2, d3) == d2) {
        inter = x1 ;
        x1 = x2 ;
        x2 = inter ;
        inter = y1 ;
        y1 = y2 ;
        y2 = inter ;
      }

      x4 = (or*x3 + x1) / (or + 1) ;
      y4 = (or*y3 + y1) / (or + 1) ;
      x5 = (or*x1 + x2) / (or + 1) ;
      y5 = (or*y1 + y2) / (or + 1) ;

      acute_triangle(x4, y4, x5, y5, x2, y2, n, g) ;
      acute_triangle(x2, y2, x3, y3, x4, y4, n, g) ;
      obtuse_triangle(x1, y1, x4, y4, x5, y5, n, g) ;
    }
    else
    {
      int px [] = {0, 0, 0} ;
      px[0] = (int)x1 ;       px[1] = (int)x2 ;       px[2] = (int)x3 ;
      int py [] = {0, 0, 0} ;
      py[0] = (int)y1 ;       py[1] = (int)y2 ;       py[2] = (int)y3 ;
      Polygon p  =  new Polygon (px, py, 3) ;

      g.setColor(bleu) ;
      g.fillPolygon(p) ;

      g.setColor(noir) ;
      g.drawPolygon(p) ;
    }
  }

  void obtuse_triangle(double x1, double y1, double x2, double y2, double x3, double y3, long n, Graphics g)  {
    double x4, y4 ;
    double d1, d2, d3 ;
    double inter ;

    if (n < niveau) {
      d1 = distance(x1, y1, x2, y2) ;
      d2 = distance(x1, y1, x3, y3) ;
      d3 = distance(x2, y2, x3, y3) ;

      n++ ;

      if (MAX(d1, d2, d3) == d1) {
        inter = x1 ;
        x1 = x3 ;
        x3 = inter ;
        inter = y1 ;
        y1 = y3 ;
        y3 = inter ;
      }
      else if (MAX(d1, d2, d3) == d2) {
        inter = x1 ;
        x1 = x2 ;
        x2 = inter ;
        inter = y1 ;
        y1 = y2 ;
        y2 = inter ;
      }

      x4 = (or*x2 + x3) / (or + 1) ;
      y4 = (or*y2 + y3) / (or + 1) ;

      acute_triangle(x1, y1, x3, y3, x4, y4, n, g) ;
      obtuse_triangle(x1, y1, x2, y2, x4, y4, n, g) ;
    }
    else
    {
      int px [] = {0, 0, 0} ;
      px[0] = (int)x1 ;       px[1] = (int)x2 ;       px[2] = (int)x3 ;
      int py [] = {0, 0, 0} ;
      py[0] = (int)y1 ;       py[1] = (int)y2 ;       py[2] = (int)y3 ;
      Polygon p  =  new Polygon (px, py, 3) ;

      g.setColor(jaune) ;
      g.fillPolygon(p) ;

      g.setColor(noir) ;
      g.drawPolygon(p) ;
    }
  }

  double distance(double x1, double y1, double x2, double y2)   {
    return Math.sqrt((x2-x1) * (x2-x1) + (y2-y1) * (y2-y1)) ;
  }

  double MIN(double x1, double x2, double x3)   {
    if ((x1 <= x2) & (x1 <= x3)) {
      return x1 ;
    }
    else if ((x2 <= x1) & (x2 <= x3)) {
      return x2 ;
    }
    else {
      return x3 ;
    }
  }

  double MAX(double x1, double x2, double x3)   {
    if ((x1 >= x2) & (x1 >= x3)) {
      return x1 ;
    }
    else if ((x2 >= x1) & (x2 >= x3)) {
      return x2 ;
    }
    else {
      return x3 ;
    }
  }

}


app/main.f90
! This file is part of the penrose_tiling Fortran project.
! Copyright (C) 2023 Vincent MAGNIN
!
! This is free software; you can redistribute it and/or modify
! it under the terms of the GNU General Public License as published by
! the Free Software Foundation; either version 3, or (at your option)
! any later version.
!
! This software is distributed in the hope that it will be useful,
! but WITHOUT ANY WARRANTY; without even the implied warranty of
! MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
! GNU General Public License for more details.
!
! You should have received a copy of the GNU General Public License along with
! this program; see the files LICENSE and LICENSE_EXCEPTION respectively.
! If not, see <http://www.gnu.org/licenses/>.
!------------------------------------------------------------------------------
! Contributed by Vincent Magnin
! GFA BASIC version: 1991-09-12
! Java version: 2002-12-14
! Fortran version: 2023-06-01
! Last modifications: 2023-06-05
!------------------------------------------------------------------------------

program main
    ! Use the cairo-fortran bindings:
    use cairo
    use cairo_enums
    ! Cairo uses C doubles:
    use, intrinsic :: iso_c_binding, only: dp=>c_double, c_ptr
    use triangles, only: phi, acute_triangle, nb_acute, nb_obtuse

    implicit none
    ! Cairo surface and Cairo context:
    type(c_ptr) :: surface, cr
    character(*), parameter :: FILENAME = "penrose_tiling.svg"
    ! Size of the surface in mm (landscape A4 paper):
    real(dp), parameter :: IMAGE_WIDTH  = 297._dp
    real(dp), parameter :: IMAGE_HEIGHT = 210._dp
    real(dp), parameter :: d = 150._dp
    real(dp), parameter :: h = d * sqrt(phi**2 - 1._dp/4._dp)
    ! Vertices of the main triangle:
    complex(dp) :: P1, P2, P3
    ! Be careful, 12 recursions will generate a 12.7 Mio SVG file, and it grows
    ! like the Fibonacci sequence...
    integer, parameter  :: MAX_RECURSION = 10

    ! The object will be rendered in a SVG file:
    surface = cairo_svg_surface_create(FILENAME//c_null_char, &
                                      & IMAGE_WIDTH, IMAGE_HEIGHT)
    call cairo_svg_surface_set_document_unit(surface, CAIRO_SVG_UNIT_MM)
    cr = cairo_create(surface)

    ! Initialize parameters:
    call cairo_set_antialias(cr, CAIRO_ANTIALIAS_BEST)
    call cairo_set_line_width(cr, 0.02_dp)

    ! We work in the complex plane:
    P1 = 0
    P2 = cmplx(0, d,   dp)
    P3 = cmplx(h, d/2, dp)

    ! Starting the recursive algorithm:
    call acute_triangle(P1, P2, P3, 1, MAX_RECURSION, cr)

    ! Finalyze:
    call cairo_destroy(cr)
    call cairo_surface_destroy(surface)

    print *, "----------------------------------"
    print *, MAX_RECURSION, "recursions"
    print *, "These are Fibonacci numbers:"
    print *, nb_acute, "blue acute triangles"
    print *, nb_obtuse, "yellow obtuse triangles"
    print *, "----------------------------------"
    print *, "Output file: ", FILENAME
    print *, "----------------------------------"
end program main
